docs/rfc/0003-collaborative-objects.md

```diff
@@ -0,0 +1,649 @@
+
+
+### Schema extensions
+
+#### `rad_signed_by`
+
+Many collaborative data structures will need to make statements about who is
+allowed to change what parts of a structure. To achieve this we extend the 
+json schema language with some custom metadata, the `rad_signed_by` property. This
+property can be placed on any `object` schema. It's value is an
+object with two keys, an array of properties which must be signed, and array of 
+radicle URNs who's signature must be present.
+
+This property implies a required `signatures` property which contains a mapping
+from URN to a description of the signature. (Maybe a JWS?).
+
+Any schema which has this property will result in some additional validation.
+Librad will encode the relevant keys of the target object using canonical JSON
+and then check that a signature over them is valid with respect to the given
+keys. 
+
+
+#### `frozen`
+
+Some attributes should never be changed, for example the ID of an issue, or a
+nonce on a comment. Any schema can add the metadata key, `frozen: true` to
+indicate that after it's initial creation any change which modifies it is
+invalid.
+
+#### `automerge_type`
+
+This can take the value `"text"` if  placed on `string` properties to indicate
+that they should be represented in an `Automerge.Text` data type.
+
+
+### Strange Perspectives
+
+This model introduces some counter-intuitive properties. For example, I might
+"create an issue" in a repository and anyone who is tracking me would see that
+issue, but people who are tracking the project but don't have me in their
+tracking graph will only see the issue if the maintainer replies to it. It's
+hard to see how you would do things like "link to an issue" under these
+constraints. This is inherent to the network model though, rather than being a
+specific problem of this architecture.
+
+We can work around some of this weirdness using seed nodes. If we consider
+seed nodes 
+
+## APIs
+
+The APIs librad will provide:
+
+- enumerate collaborative objects of a particular type
+- retrieve an object with a particular ID as a JSON representation for
+  applications which only wish to read data
+- retrieve an object with a particular ID as an Automerge document for
+  applications which wish to write data
+- update an object by providing the bytes of an automerge change which updates
+  the document
+- create a new object from a JSON object, a JSON schema, and a type name
+  
+Note that I am referring to "the binary representation of an automerge x" 
+because the automerge API works in terms of binary changes.
+
+This new api will live in a new top level module at
+`librad::collaborative_objects`. An initial sketch looks like this:
+
+```rust
+struct CollaborativeObjectStore {
+    storage: git::storage::Pool,
+    signer: signer::Signer,
+}
+
+enum History {
+    Automerge(Vec<Vec<u8>>)
+}
+
+struct ObjectId(String);
+struct TypeName(String);
+struct Schema(..);
```

Comment:
`..`? :eyes: Is this also just bytes or does the `..` mean you're still figuring that out?

Reply:
I'm still working it out :slightly_smiling_face:. More to the point, it doesn't actually matter for the API as the purpose of the tuple struct is to hide the implementation, I guess I could do the same for the `ObjectId` and `TypeName` for consistency.